cpp/src/arrow/ipc/reader.cc

```diff
@@ -2052,6 +2052,39 @@ Status Listener::OnRecordBatchWithMetadataDecoded(
   return OnRecordBatchDecoded(std::move(record_batch_with_metadata.batch));
 }
 
+namespace {
+Status CopyArrayData(std::shared_ptr<ArrayData> data) {
```

Comment:
Maybe, we should not do this...

Reply:
Would `ArrayData*` be better in this place?

Reply:
Good catch!
It will reduce reference count related cost.
I'll do it after we get a consensus with this copy approach is reasonable. (I hope that we have better approach...)